packages/labs/context/src/lib/controllers/context-consumer.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {ContextRequestEvent} from '../context-request-event.js';
+import {ContextKey, ContextType} from '../context-key.js';
+import {ReactiveController, ReactiveElement} from 'lit';
+
+export type DispatchOptions = 'connected' | 'updated' | 'raf' | 'microtask';
+
+/**
+ * ContextConsumer is a ReactiveController which binds a custom-element's
+ * lifecycle to the Context API. When an element is connected to the DOM it
+ * will emit the context-request event, invoking the callback set on the
+ * controller when the context request is satisfied. It will also call
+ * the dispose method provided by the Context API when the element is
+ * disconnected.
+ */
+export class ContextConsumer<
+  Context extends ContextKey<unknown, unknown>,
+  HostElement extends ReactiveElement
+> implements ReactiveController
+{
+  private provided = false;
+
+  constructor(
+    protected host: HostElement,
+    private context: Context,
+    private callback: (
+      value: ContextType<Context>,
+      dispose?: () => void
+    ) => void,
+    private dispatchOn: DispatchOptions = 'connected',
+    private subscribe: boolean = false
+  ) {
+    this.host.addController(this);
+  }
+
+  private unsubscribe?: () => void;
+
+  hostConnected(): void {
+    this.scheduleRequest(this.dispatchOn === 'connected');
+  }
+  hostDisconnected(): void {
+    if (this.unsubscribe) {
+      this.unsubscribe();
+      this.unsubscribe = undefined;
+    }
+  }
+  hostUpdated(): void {
+    if (!this.provided && this.dispatchOn === 'updated') {
+      this.scheduleRequest(true);
+    }
+  }
+
+  private scheduleRequest(immediate: boolean) {
+    if (immediate) {
+      this.dispatchRequest();
+    } else if (this.dispatchOn === 'raf') {
+      requestAnimationFrame(() => this.dispatchRequest());
+    } else if (this.dispatchOn === 'microtask') {
+      queueMicrotask(() => this.dispatchRequest());
+    }
+  }
+
+  private dispatchRequest() {
+    this.host.dispatchEvent(
+      new ContextRequestEvent(
+        this.context,
+        (value, unsubscribe) => {
+          // some providers will pass an unsubscribe function indicating they may provide future values
+          if (this.unsubscribe) {
+            // if the unsubscribe function changes this implies we have changed provider
+            if (this.unsubscribe !== unsubscribe) {
+              // cleanup the old provider
+              this.provided = false;
+              this.unsubscribe();
+            }
+            // if we don't support subscription, immediately unsubscribe
+            if (!this.subscribe) {
+              this.unsubscribe();
+            }
+          }
+
+          // only invoke callback if we are either expecting updates or have not yet
+          // been provided a value
+          if (!this.provided || this.subscribe) {
```

Comment:
Is this called if `subscribe` is false?

Reply:
Yes, a consumer can request a context without subscribing for subsequent updates (`subscribe` is false). The second condition here caters for the 'multiple updates' case where a consumer 'subscribes' to the context for repeated update.

Reply:
IMO the option to have a single context (updates consumer once) as specified in protocol is a mistake. The provided value may be not defined yet or be lazy loaded at the time the context event occurs. And requiring to the consumer know when is lazy loaded or not, creates a dependency to an implementation detail.

Also complicates the things a bit

The consumer, if only wants the first value, should call the unsubscribe at the time receives the first meaningful value

Reply:
I agree to some extent, defaulting to subscription rather than one-time request would simplify implementation a little. Though it would also have a potential overhead for all those context requests which are one-time only since the provider needs to create the unsubscribe function and it needs to be called. So you pay for that implementation simplicity with runtime overhead?

Reply:
Is the runtime overhead measurable?

Anyway, more important than implementation simplification is the decoupling from provider. 

Say you have a provider for `model` state. Initially is implemented as a sigleton / available from start.
Consumers are implemented with `multiple: false` 

Later the provider changes its implementation to lazy load `model`, sending the initial value as undefined. You need to change consumers implementation.

It's bad when provider and consumer are developed in same project. Now think when provider/consumer are developed by third party.

Reply:
Performance testing was going to be something the lit team would follow up with after getting this initial implementation merged. Might be worth exploring then.

I get your point here though. There is certainly some simplicity in only having the single model of subscription behavior. I'm open to changing it but we would need to get agreement from the community protocol group on that change. That might be easier to do so after we actually have this implementation published somewhere for people to try out.

You might also argue that a provider that switched to a lazy loaded value from an immediately provided value is breaking its initial contract, and that should really be done with a new context key?

Reply:
Filed https://github.com/lit/lit/issues/2762 to track adding performance tests.